streams/src/main/java/org/apache/kafka/streams/kstream/Suppress.java

```diff
@@ -0,0 +1,90 @@
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressImpl.BufferConfigImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressImpl.IntermediateSuppressionImpl;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import java.time.Duration;
+
+@SuppressWarnings("unused") // complaint about the top-level generic parameters, which actually are necessary
+public interface Suppress<K, V> {
+    interface TimeDefinition<K, V> {
+        long time(ProcessorContext context, K k, V v);
+    }
+
+    enum BufferFullStrategy {
+        EMIT,
+        SPILL_TO_DISK,
+        SHUT_DOWN
+    }
+
+    interface BufferConfig<K, V> {
+        static <K, V> BufferConfig<K, V> withBufferKeys(final long numberOfKeysToRemember) {
+            final BufferConfig<K, V> bufferConfig = new BufferConfigImpl<>();
+            return bufferConfig.bufferKeys(numberOfKeysToRemember);
+        }
+
+        BufferConfig<K, V> bufferKeys(final long numberOfKeysToRemember);
+
+        static <K, V> BufferConfig<K, V> withBufferBytes(final long bytesToUseForSuppressionStorage, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
+            final BufferConfig<K, V> bufferConfig = new BufferConfigImpl<>();
+            return bufferConfig.bufferBytes(bytesToUseForSuppressionStorage, keySerializer, valueSerializer);
+        }
+
+        BufferConfig<K, V> bufferBytes(final long bytesToUseForSuppressionStorage, final Serializer<K> keySerializer, final Serializer<V> valueSerializer);
+
+        static <K, V> BufferConfig<K, V> withBufferFullStrategy(final BufferFullStrategy bufferFullStrategy) {
+            return new BufferConfigImpl<K, V>().bufferFullStrategy(bufferFullStrategy);
+        }
+
+        BufferConfig<K, V> bufferFullStrategy(final BufferFullStrategy bufferFullStrategy);
+    }
+
+    interface IntermediateSuppression<K, V> {
+        static <K, V> IntermediateSuppression<K, V> withEmitAfter(final Duration timeToWaitForMoreEvents) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.emitAfter(timeToWaitForMoreEvents);
+        }
+
+        IntermediateSuppression<K, V> emitAfter(final Duration timeToWaitForMoreEvents);
+
+        static <K, V> IntermediateSuppression<K, V> withBufferConfig(final BufferConfig<K, V> bufferConfig) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.bufferConfig(bufferConfig);
+        }
+
+        IntermediateSuppression<K, V> bufferConfig(final BufferConfig<K, V> bufferConfig);
+    }
+
+    static <K extends Windowed, V> Suppress<K, V> emitFinalResultsOnly(final BufferConfig<K, V> bufferConfig) {
+        if (((BufferConfigImpl<K, V>) bufferConfig).getBufferFullStrategy() == BufferFullStrategy.EMIT) {
```

Comment:
Hmm.. this makes me thinking if we should consider duplicate the `BufferFullStrategy` to `FinalBufferFullStrategy` and `IntermediateBufferFullStrategy` and also the corresponding caller `BufferConfig` as well, to replace runtime error with complication error?

Reply:
Yeah, I was thinking something similar... I'll sketch something up.